internal/pkg/api/handleEnroll.go

```diff
@@ -252,6 +260,105 @@ func (et *EnrollerT) _enroll(
 				Msg("Error when trying to delete old agent with enrollment id")
 			return nil, err
 		}
+		// deleted, so clear the ID so code below knows it needs to be created
+		agent.Id = ""
+	}
+
+	var agentID string
+	if req.Id != nil && *req.Id != "" {
+		agentID = *req.Id
+
+		// Possible this agent already exists.
+		vSpan, vCtx := apm.StartSpan(ctx, "checkAgentID", "validate")
+		var err error
+		agent, err = dl.FindAgent(vCtx, et.bulker, dl.QueryAgentByID, dl.FieldID, agentID)
+		if err != nil {
+			zlog.Debug().Err(err).
+				Str("ID", agentID).
+				Msg("Agent with ID not found")
+			if !errors.Is(err, dl.ErrNotFound) && !strings.Contains(err.Error(), "no such index") {
+				vSpan.End()
+				return nil, err
+			}
+		} else if !agent.Active {
+			// inactive agent has been unenrolled and the API key has already been invalidated
+			// delete the current record as the new enrollment will overwrite this one
+			zlog.Debug().
+				Str("ID", agentID).
+				Msg("Inactive agent with ID found")
+			err = deleteAgent(ctx, zlog, et.bulker, agent.Id)
+			if err != nil {
+				zlog.Error().Err(err).
+					Str("AgentId", agent.Id).
+					Msg("Error when trying to delete old agent with same id")
+				return nil, err
+			}
+			// deleted, so clear the ID so code below knows it needs to be created
+			agent.Id = ""
+		} else {
+			zlog.Debug().
+				Str("ID", agentID).
+				Msg("Active agent with ID found")
+		}
+		vSpan.End()
+
+		if agent.Id != "" {
+			// confirm that this agent has a set replace token
+			// one is required or replacement of this already enrolled and active
+			// agent is not allowed
+			if agent.ReplaceToken == "" {
+				zlog.Warn().
+					Str("AgentId", agent.Id).
+					Msg("Existing agent with same ID already enrolled without a replace token set")
+				return nil, ErrAgentNotReplaceable
+			}
+			if req.ReplaceToken == nil || *req.ReplaceToken == "" {
+				zlog.Warn().
+					Str("AgentId", agent.Id).
+					Msg("Existing agent with same ID already enrolled; no replace token given during enrollment")
+				return nil, ErrAgentNotReplaceable
+			}
+			err = bcrypt.CompareHashAndPassword([]byte(agent.ReplaceToken), []byte(*req.ReplaceToken))
+			if err != nil {
+				// not the same, cannot replace
+				// provides no real reason as that would expose to much information
+				return nil, ErrAgentNotReplaceable
+			}
+
+			// confirm that its on the same policy
+			// it is not supported to have it the same ID enroll into different policies
+			if agent.PolicyID != policyID {
+				zlog.Warn().
+					Str("AgentId", agent.Id).
+					Str("PolicyId", policyID).
+					Str("CurrentPolicyId", agent.PolicyID).
+					Msg("Existing agent with same ID already enrolled into another policy")
+				return nil, ErrAgentInAnotherPolicy
```

Comment:
very minor: i wonder if this is not telling too much, attacker could try agentID and token. this would tell token is correct, policy is not. Maybe returning ErrAgentNotReplaceable with log unchanged would make more sense?